tools/cwtool/TxSimulator.hs

```diff
@@ -2,113 +2,242 @@
+      liftIO (spvSim sc hdr pwo) >>= assign psSpvSupport
+      _r <- trace (logFunction cwLogger) "execBlock" () 1 $
+          execBlock hdr (payloadWithOutputsToPayloadData pwo) pde'
+      liftIO $ _cpSave cp (_blockHash hdr)
+      doBlock False hdr rest
 
+-- | Block-scoped SPV mock by matching cont proofs to payload txs.
+-- Transactions are eliminated by searching for first matching proof in input;
+-- there should always be as many exact matches as proofs.
+-- TODO Heuristic for failed SPV needs to interrogate failures from outputs.
+spvSim :: SimConfig -> BlockHeader -> PayloadWithOutputs -> IO SPVSupport
+spvSim sc bh pwo = do
+  mv <- newMVar (V.toList (_payloadWithOutputsTransactions pwo))
+  return $ SPVSupport (_spvSupport noSPVSupport) (go mv)
+  where
+    go mv cp = modifyMVar mv $ searchOuts cp
+    searchOuts _ [] = return ([],Left "spv: proof not found")
+    searchOuts cp@(ContProof pf) ((Transaction ti,TransactionOutput o):txs) =
+      case codecDecode (chainwebPayloadCodec (Just (scVersion sc,_blockHeight bh))) ti of
+        Left {} -> internalError "input decode failed"
+        Right cmd -> case _pPayload $ payloadObj $ _cmdPayload cmd of
+          Continuation cm | _cmProof cm == Just cp -> do
+            cr :: CommandResult Hash <- decodeStrictOrThrow o
+            case _pactResult (_crResult cr) of
+              _ -> do -- TODO failed SPV error heuristic
+                -- the following adapted from Chainweb.Pact.SPV.verifyCont with matching errors
+                t <- decodeB64UrlNoPaddingText $ T.decodeUtf8 pf
+                case decodeStrict' t of
+                  Nothing -> internalError "unable to decode continuation proof"
+                  Just (TransactionOutputProof pcid p :: TransactionOutputProof SHA512t_256) -> do
+                    unless (pcid == scChain sc) $
+                      internalError "cannot redeem continuation proof on wrong target chain"
+                    TransactionOutput tout <- proofSubject p
+                    case decodeStrict' tout :: Maybe (CommandResult Hash) of
+                      Nothing -> internalError "unable to decode spv transaction output"
+                      Just cro -> case _crContinuation cro of
+                        Nothing -> return (txs,Left "no pact exec found in command result")
+                        Just pe -> return (txs,Right pe)
+          _ -> searchOuts cp txs
+
+setupClient :: SimConfig -> IO ClientEnv
+setupClient sc = flip mkClientEnv (scApiHostUrl sc) <$> newTlsManagerWith mgrSettings
+  where
+    mgrSettings = mkManagerSettings
+        (TLSSettingsSimple True False False)
+        Nothing
+
+-- | note, fetches [low - 1, hi] to have parent headers
+fetchHeaders :: SimConfig -> ClientEnv -> IO [BlockHeader]
+fetchHeaders sc cenv = do
+  r <- (`runClientM` cenv) $
+      headersClient (scVersion sc) (scChain sc) Nothing Nothing
+      (Just $ fromIntegral $ pred $ fst $ scRange sc)
```

Comment:
why `pred` here ... is it to make sure we always fetch the parent header? If so, why not make that part of the range specification (or just comment it)?

Reply:
it was already commented :)